### PR TITLE
hammer: RGW shouldn't send Content-Type nor Content-Length for 304 responses

### PR DESCRIPTION
**Pull request by Vicente-Cheng**
http://tracker.ceph.com/issues/15283

### CODE REVIEW

**Comment by tchaikov** [comment]
just not sure if we add the notes in the cherry-picked commits
```
Fixes: #14005
Backport: hammer

Signed-off-by: Vicente Cheng <freeze.bilsted@gmail.com>
```

**Comment by Vicente-Cheng** [comment]
@tchaikov Yes, I am not sure, too.
or just add backport information is enough?

**Comment by dachary** [comment]
The conflict resolution comment matches the difference

```
commit=19dbc2598d15d0676017abe4fb75f7ddc3248119 ; picked_from=$(git show --no-patch --pretty=%b $commit  | perl -ne 'print if(s/.*cherry picked from commit (\w+).*/$1/)') ; diff -u --ignore-matching-lines '^[^+-]'   <(git show $picked_from) <(git show $commit)
--- /dev/fd/63	2016-03-31 11:37:30.911948034 +0200
+++ /dev/fd/62	2016-03-31 11:37:30.927947932 +0200
@@ -149,11 +154,12 @@
 +  dump_status(s, s->err.http_ret, http_status_names[s->err.http_ret]);
  }
  
- void dump_errno(struct req_state *s, int http_ret)
+-void dump_errno(struct req_state *s, int err)
++void dump_errno(struct req_state *s, int http_ret)
  {
 -  char buf[32];
--  snprintf(buf, sizeof(buf), "%d", http_ret);
--  dump_status(s, buf, http_status_names[http_ret]);
+-  snprintf(buf, sizeof(buf), "%d", err);
+-  dump_status(s, buf, http_status_names[s->err.http_ret]);
 +  dump_status(s, http_ret, http_status_names[http_ret]);
  }
```

File: src/rgw/rgw_rest_s3.cc [hunk 1]
```diff
@@ -167,10 +167,14 @@ int RGWGetObj_ObjStore_S3::send_response_data(bufferlist& bl, off_t bl_ofs, off_
     s->cio->print("%s: %s\r\n", riter->first.c_str(), riter->second.c_str());
   }
 
-  if (!content_type)
-    content_type = "binary/octet-stream";
+  if (ret == ERR_NOT_MODIFIED) {
```
Comment (smunaut) [inline src/rgw/rgw_rest_s3.cc]: Pretty sure this should be ret == -ERR_NOT_MODIFIED

Comment (dachary) [inline src/rgw/rgw_rest_s3.cc]: Would you mind creating an issue at http://tracker.ceph.com/projects/rgw/issues/new for that ? It's in master and should be fixed there.

Comment (smithfarm) [inline src/rgw/rgw_rest_s3.cc]: @dachary Issue is http://tracker.ceph.com/issues/16327, fixed by #9725

**Comment by smithfarm** [comment]
@Vicente-Cheng Please cherry-pick fc38346 into this PR (see http://tracker.ceph.com/issues/16327)

**Comment by Vicente-Cheng** [comment]
@smithfarm 
hi Nathan,

sorry about late update.
already done this cherry-pick commit.
retest it please.

thanks

**Comment by smithfarm** [comment]
This PR passed an rgw suite with some valgrind-related failures that appear harmless - see http://tracker.ceph.com/issues/15895#note-19 

Do you think this PR is ready to merge?

**Comment by oritwas** [comment]
lgtm

**Pull request merged**